File: online judge.kaner murillo/c++/1216.cpp
//https://onlinejudge.org/external/12/1216.pdf
//1216

#include<cstdio>
#include<iostream>
#include<vector>
#include<map>
#include<climits>
#include<cmath>
#include<queue>
#include<algorithm>

using namespace std;

int total;

void prim(vector<pair<int, int>>&graph, vector<int>&dist, vector<int>&pre){

    int i, node, weight, auxWeight;

    for(i = 0; i<dist.size(); i++){

        dist[i] = INT_MAX;

    }

    dist[0] = 0;
    vector<int>visits(graph.size(), 0);
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>>q;

    q.push(make_pair(0, 0));

    while(!q.empty()){

        weight = q.top().first;
        node = q.top().second;
        
        visits[node] = 1;

        q.pop();

        if(weight == dist[node]){
            
            total+= weight;

            for(i = 0; i<graph.size(); i++){

                if(i != node){

                    auxWeight = ceil(sqrt(pow(graph[i].first - graph[node].first, 2) + pow(graph[i].second - graph[node].second, 2)));

                    if(!visits[i] && auxWeight < dist[i]){

                        dist[i] = auxWeight;
                        pre[i] = node;
                        q.push(make_pair(auxWeight, i));
                    }
                }
            }
        }
    }
}

bool comp(int &a, int &b){

    return a > b;
}

int main(){

    int cases, receiver, x, y, i;

    scanf("%i", &cases);

    while(cases--){

        scanf("%i", &receiver);
        vector<pair<int, int>>graph;
        total = 0;

        while(scanf("%i", &x) && x != -1){

            scanf("%i", &y);

            graph.push_back(make_pair(x, y));
        }

        vector<int>dist(graph.size());
        vector<int>pre(graph.size(), -1);

        prim(graph, dist, pre);

        sort(dist.begin(), dist.end(), comp);

        receiver--;
        
        printf("%i\n", dist[receiver]);

    }

    return 0;
}

File: online judge.kaner murillo/python/1216.py
#https://onlinejudge.org/external/12/1216.pdf
#1216

from sys import stdin
from heapq import heappop, heappush
from math import ceil, sqrt

def prim(graph):

    dist = [float('inf') for i in range(len(graph))]
    visits = [0 for i in range(len(graph))]
    q = [(0, 0)]
    dist[0] = 0

    while len(q):

        point = heappop(q) 
        visits[point[1]] = 1

        if dist[point[1]] == point[0]:
            for i in range(len(graph)):
                if i != point[1]:
                    weight = ceil(sqrt(pow(graph[point[1]][0]-graph[i][0], 2) + pow(graph[point[1]][1]-graph[i][1], 2)))
                    if not visits[i] and weight < dist[i]:
                        dist[i] = weight
                        heappush(q, (weight, i))

    return dist
                  
def main():

    cases = int(stdin.readline())

    for i in range(cases):

        receiver = int(stdin.readline())
        receiver-=1
        points = []
        point = [int(x) for x in stdin.readline().split()]

        while point[0] != -1:
            points.append(point)
            point = [int(x) for x in stdin.readline().split()]
        
        dist = prim(points)
        dist.sort(key=lambda x: -x)
        print(dist[receiver])
        
main()